client/client.go

```diff
@@ -176,37 +173,84 @@ func (c *Client) Run(ctx context.Context) error {
 	// https://jira.percona.com/browse/PMM-4245
 
 	oneDone := make(chan struct{}, 5)
-	go func() {
-		c.jobsRunner.Run(ctx)
-		oneDone <- struct{}{}
-	}()
-	go func() {
-		c.processActionResults()
-		oneDone <- struct{}{}
-	}()
-	go func() {
-		c.processJobsResults()
-		oneDone <- struct{}{}
-	}()
-	go func() {
-		c.processSupervisorRequests()
-		oneDone <- struct{}{}
-	}()
-	go func() {
-		c.processChannelRequests(ctx)
-		oneDone <- struct{}{}
-	}()
 
-	<-oneDone
-	go func() {
-		<-oneDone
-		<-oneDone
-		<-oneDone
-		<-oneDone
-		c.l.Info("Done.")
-		close(c.done)
-	}()
-	return nil
+	doConnect := make(chan struct{}, 1)
+	doConnect <- struct{}{}
+	var err error
+
+	for {
```

Comment:
this doesn't look thread-safe.
High possibility of running a few `process` go routines.